File: solutions/javascript/src/structure/dynamic/DynamicBinaryTreeADT.js
class BinaryTreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class DynamicBinaryTreeADT {
  constructor() {
    this.root = null;
  }

  getRoot() {
    return this.root ? this.root.value : null;
  }

  getLeft() {
    return this.root ? this.root.left : null;
  }

  getRight() {
    return this.root ? this.root.right : null;
  }

  isEmpty() {
    return this.root === null;
  }

  add(value) {
    this.root = this.#addRecursive(this.root, value);
  }

  #addRecursive(node, value) {
    if (node === null) return new BinaryTreeNode(value);

    if (value < node.value) {
      node.left = this.#addRecursive(node.left, value);
    } else {
      node.right = this.#addRecursive(node.right, value);
    }
    return node;
  }

  remove(value) {
    this.root = this.#removeRecursive(this.root, value);
  }

  #removeRecursive(node, value) {
    if (node === null) return null;

    if (value < node.value) {
      node.left = this.#removeRecursive(node.left, value);
    } else if (value > node.value) {
      node.right = this.#removeRecursive(node.right, value);
    } else {
      // Nodo con un solo hijo o sin hijos
      if (node.left === null) return node.right;
      if (node.right === null) return node.left;

      // Nodo con dos hijos: obtener el mínimo del subárbol derecho
      let minLargerNode = this.#getMin(node.right);
      node.value = minLargerNode.value;
      node.right = this.#removeRecursive(node.right, minLargerNode.value);
    }
    return node;
  }

  #getMin(node) {
    while (node.left !== null) node = node.left;
    return node;
  }
}

File: solutions/python/structure/dynamic/DynamicBinaryTreeADT.py
class DynamicBinaryTreeADT:
    def __init__(self, value=None):
        self.root = value
        self.left = None
        self.right = None

    def getRoot(self):
        return self.root

    def getLeft(self):
        return self.left

    def getRight(self):
        return self.right

    def add(self, value):
        if self.root is None:
            self.root = value
        elif value < self.root:
            if self.left is None:
                self.left = BinaryTreeADT(value)
            else:
                self.left.add(value)
        else:
            if self.right is None:
                self.right = BinaryTreeADT(value)
            else:
                self.right.add(value)

    def remove(self, value):
        # Implementación básica para eliminar nodos de un BST
        if self.root is None:
            return self

        if value < self.root:
            if self.left:
                self.left = self.left.remove(value)
        elif value > self.root:
            if self.right:
                self.right = self.right.remove(value)
        else:
            if self.left is None:
                return self.right
            if self.right is None:
                return self.left
            # Nodo con dos hijos: reemplazar con el menor de los mayores
            min_larger_node = self.right
            while min_larger_node.left:
                min_larger_node = min_larger_node.left
            self.root = min_larger_node.root
            self.right = self.right.remove(min_larger_node.root)
        return self

    def isEmpty(self):
        return self.root is None
